Clean headers from zup

Myoldmopar opened the pull request:
Pull request overview
---------------------
Header cleanups from Z upwards.  This work started several weeks ago, and I just polished it up to get it merged in while I'm working on other enum related changes.  I'll give a walkthrough to discuss the types of changes included here, but to merge I would expect this to be totally green on CI, no impact on output and minimal/no impact on runtime.  However a huge improvement on code maintenance and readability.

**NOTE: ENHANCEMENTS MUST FOLLOW A SUBMISSION PROCESS INCLUDING A FEATURE PROPOSAL AND DESIGN DOCUMENT PRIOR TO SUBMITTING CODE**

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [ ] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

Myoldmopar committed:
Clean up ZoneTempPredictorCorrector header

Myoldmopar committed:
Cleanup zoneplenum header

Myoldmopar committed:
Clean ZoneEquipmentManager header

Myoldmopar committed:
Clean ZoneDehumidifier header

Myoldmopar committed:
Clean ZoneContamPredCorr header

Myoldmopar committed:
Clean ZoneAirLoopEquip header

Myoldmopar committed:
Fix header output bug

Myoldmopar committed:
Clean windturbine header

Myoldmopar committed:
Clean WindowManager header - incremental

Myoldmopar committed:
Cleaning up window manager header blech

Myoldmopar committed:
Fix passing array search length into Interpolate

Myoldmopar committed:
Cleanup unused variables and comment lines in header

Myoldmopar committed:
Refactor numpt to std::array

Myoldmopar committed:
Converted three more Array1D to std::array

Myoldmopar committed:
Refactored a couple more Array1D to std::array

Myoldmopar committed:
Converted two more Array2D to std::arrays

Myoldmopar committed:
Converted three more Array2D to std::arrays

Myoldmopar committed:
Converted 4 more Arrays to arrays

Myoldmopar committed:
Clang format

Myoldmopar committed:
Converted three more Array1D to std::array

Myoldmopar committed:
Converted 4 more Array1D to std::array

Myoldmopar committed:
Clang format

Myoldmopar committed:
Remove 10 Arrays from state that arent even being used

Myoldmopar committed:
Removal and conversion of more Array to std::array

Myoldmopar committed:
Quick convert an Array2D to std::array

Myoldmopar committed:
Revert "Quick convert an Array2D to std::array"

This reverts commit 0b54719421e20652a44a81953b6761f8f562144e.

Myoldmopar committed:
Convert 10ish Array1D to std::array

Myoldmopar committed:
Merge develop; resolve conflicts

Myoldmopar committed:
Clang format

amirroth commented on src/EnergyPlus/ConvectionCoefficients.hh:
```diff
@@ -236,6 +236,7 @@ namespace ConvectionCoefficients {
         // Default Constructor
         RoofGeoCharacteristicsStruct() : BoundSurf(4)
         {
+            this->BoundSurf = Array1D<Vector>(4);
```
This can be a `std::array<Vector, 4>` with no explicit construction here.

Myoldmopar replied:
Yeah I was just looking at why I even made that change...I'll fix it up.

Myoldmopar replied:
If we make this change to std::array, it requires changes the signature for CreateNewellAreaVector, which then hits a bunch of places in SurfaceGeometry, which will lead to more changes.  I can certainly go down that path, but it will much more likely start conflicting other branches.  For right now I'll hold off but if you feel it's a critical change then I'll be happy to do it here.

amirroth replied:
Leave it for now.

amirroth commented on src/EnergyPlus/WindTurbine.cc:
```diff
@@ -240,11 +241,11 @@ namespace WindTurbine {
+        state.dataWindTurbine->WindTurbineSys.allocate(NumWindTurbines);
 
-        for (WindTurbineNum = 1; WindTurbineNum <= state.dataWindTurbine->NumWindTurbines; ++WindTurbineNum) {
+        for (WindTurbineNum = 1; WindTurbineNum <= NumWindTurbines; ++WindTurbineNum) {
```
Reduce induction variable scope.

Myoldmopar replied:
Sure, the only reason I didn't automatically do that is sometimes these counters are used in GetInput routines for multiple components that might be stored in the same array.  It's rare, but it happens around E+.  So if you localize them, then odd issues and hard to find diffs slow down the review process.  I will check if I can clean this specific one up.

amirroth commented on src/EnergyPlus/WindTurbine.hh:
```diff
@@ -85,64 +85,52 @@ namespace WindTurbine {
+        Real64 ChordArea = 0.0;                         // Chord area of a single blade for VAWTs in m2
+        Real64 DragCoeff = 0.0;                         // Empirical blade drag coefficient for VAWTs
+        Real64 LiftCoeff = 0.0;                         // Empirical blade lift coefficient for VAWTs
+        Real64 PowerCoeffC1 = 0.0;                      // Empirical power coefficient 1 for analytical calculation
```
`std::array`?

Myoldmopar replied:
Done.

amirroth commented on src/EnergyPlus/WindowManager.cc:
```diff
@@ -508,7 +508,7 @@ namespace WindowManager {
                 }
 
-                state.dataWindowManager->LayerNum(IGlass) = LayNum;
+                state.dataWindowManager->LayerNum[IGlass - 1] = LayNum;
```
Probably not easy to do immediately, but why not just make `IGlass` go from `0` to `NGlass - 1` and eliminate `IGlass - 1`?

Myoldmopar replied:
To avoid accidentally causing index issues in other arrays that we haven't touched yet, it was safer to just tweak the lookups in arrays that I am touching.  The WindowManager code has so very many arrays that I thought it better to wait until I was for sure done with all of them and then do a smart replacement in these loops separately.  That way strange diffs and a deep investigation didn't bog down this branch.

amirroth replied:
Roger Victor.

amirroth commented on src/EnergyPlus/WindowManager.cc:
```diff
@@ -547,14 +547,14 @@ namespace WindowManager {
-                        SolarSprectrumAverage(state, state.dataWindowManager->t, state.dataWindowManager->tmpTrans);
-                        SolarSprectrumAverage(state, state.dataWindowManager->rff, state.dataWindowManager->tmpReflectSolBeamFront);
-                        SolarSprectrumAverage(state, state.dataWindowManager->rbb, state.dataWindowManager->tmpReflectSolBeamBack);
+                        state.dataWindowManager->tmpTrans = solarSpectrumAverage(state, state.dataWindowManager->t[0]);
```
:+1:

amirroth commented on src/EnergyPlus/WindowManager.cc:
```diff
@@ -659,42 +659,42 @@ namespace WindowManager {
-                        rfBareVisPhi(IGlass, IPhi) = state.dataWindowManager->rfPhi(IGlass, 2);
-                        rbBareSolPhi(IGlass, IPhi) = state.dataWindowManager->rbPhi(IGlass, 1);
-                        rbBareVisPhi(IGlass, IPhi) = state.dataWindowManager->rbPhi(IGlass, 2);
+                        tBareSolPhi(IGlass, IPhi) = state.dataWindowManager->tPhi[IGlass - 1][0];
```
I would not make this comment here except for the fact that you are already messing around with the indices. I increasingly think that in cases like this you are better off having a single `array<array<struct>>` rather than three separate `array<array<Real64>>`.

amirroth commented on src/EnergyPlus/WindowManager.cc:
```diff
@@ -708,9 +708,9 @@ namespace WindowManager {
-                SolarSprectrumAverage(state, state.dataWindowManager->stPhi, state.dataWindowManager->tsolPhi(IPhi));
-                SolarSprectrumAverage(state, state.dataWindowManager->srfPhi, state.dataWindowManager->rfsolPhi(IPhi));
-                SolarSprectrumAverage(state, state.dataWindowManager->srbPhi, state.dataWindowManager->rbsolPhi(IPhi));
+                state.dataWindowManager->tsolPhi(IPhi) = solarSpectrumAverage(state, state.dataWindowManager->stPhi);
```
Same here.

amirroth commented on src/EnergyPlus/WindowManager.cc:
```diff
@@ -855,36 +855,36 @@ namespace WindowManager {
                     for (ILam = 1; ILam <= numptDAT; ++ILam) {
-                        state.dataWindowManager->wlt(IGlass, ILam) = state.dataHeatBal->SpectralData(SpecDataNum).WaveLength(ILam);
-                        state.dataWindowManager->t(IGlass, ILam) = state.dataHeatBal->SpectralData(SpecDataNum).Trans(ILam);
+                        state.dataWindowManager->wlt[IGlass - 1][ILam - 1] = state.dataHeatBal->SpectralData(SpecDataNum).WaveLength(ILam);
```
Same here.

amirroth commented on src/EnergyPlus/WindowManager.cc:
```diff
@@ -2886,144 +2866,144 @@ namespace WindowManager {
-        state.dataWindowManager->fvec = 0.0;
+        state.dataWindowManager->fvec = {0.0};
 
         {
```
I pity the person who wrote this code. For real.

amirroth commented on src/EnergyPlus/WindowManager.hh:
```diff
@@ -74,312 +74,230 @@ namespace WindowManager {
+                     gsl::span<Real64> y, // Array of data points for dependent variable
+                     int npts,            // Number of data pairs
+                     Real64 xin,          // Given value of x
+                     Real64 &yout         // Interpolated value of y at xin
```
Just return this bad boy.

Myoldmopar replied:
Done.

amirroth commented on src/EnergyPlus/WindowManager.hh:
```diff
@@ -74,312 +74,230 @@ namespace WindowManager {
+                           Real64 tf0,               // Transmittance at zero incidence angle
+                           Real64 rf0,               // Front reflectance at zero incidence angle
+                           Real64 rb0,               // Back reflectance at zero incidence angle
+                           Real64 &tfp,              // Transmittance at cs
```
Consider using a `struct` with `tf`, `rf`, and `rb` to reduce parameter passing.

Myoldmopar replied:
Given the huge number of ArrayND objects I was able to convert to std::array and achieved no-diffs, I'd like to hold off from this on this branch.  There were a couple times where I started converting a chunk of Arrays to arrays and ended up getting diffs that I ultimately reverted to avoid getting bogged down.  Assuming this comes back clean on CI once I've addressed your primary comments, I'd like to get this much in and then come back to this struct comment as well as the array of array of struct comment.  What do you think?

amirroth replied:
💯

amirroth commented on src/EnergyPlus/WindowManager.hh:
```diff
@@ -388,213 +306,202 @@ namespace WindowManager {
+    static int constexpr numt3 = 81;                                   // Number of wavelength values in the photopic response
+
+    //                                      Dens  dDens/dT  Con    dCon/dT   Vis    dVis/dT Prandtl dPrandtl/dT
+    std::array<Real64, 8> const AirProps = {1.29, -0.4e-2, 2.41e-2, 7.6e-5, 1.73e-5, 1.0e-7, 0.72, 1.8e-3};
```
There should be some enumeration for this where `DENS = 0`, `dDENSdT = 1`, etc.

Myoldmopar replied:
Yeah...or this shouldn't be an array.  I don't see what an ordered array does for these disparate physical properties.

amirroth replied:
This feels more like a `struct` than an `array`.

amirroth commented on src/EnergyPlus/WindowManager.hh:
```diff
@@ -388,213 +306,202 @@ namespace WindowManager {
+    Real64 Outir = 0.0;                                               // IR radiance of window's exterior surround (W/m2)
+    Real64 Rmir = 0.0;                                                // IR radiance of window's interior surround (W/m2)
+    Real64 Rtot = 0.0;                                                // Total thermal resistance of window (m2-K/W)
+    std::array<std::array<std::array<Real64, 5>, 5>, 3> gcon = {0.0}; // Gas thermal conductivity coefficients for each gap
```
The numbers `5` and `3` should ideally be symbolic.

amirroth commented on src/EnergyPlus/WindowManager.hh:
```diff
@@ -388,213 +306,202 @@ namespace WindowManager {
+    Real64 A45 = 0.0;
+    Real64 A67 = 0.0;
+
+    static int constexpr MaxNumOfIncidentAngles = 20;
```
`Of` 😡

amirroth commented on src/EnergyPlus/WindowManager.hh:
```diff
@@ -388,213 +306,202 @@ namespace WindowManager {
-    Array2D<Real64> top;            // Transmittance matrix for subr. op
-    Array2D<Real64> rfop;           // Front reflectance matrix for subr. op
-    Array2D<Real64> rbop;           // Back transmittance matrix for subr. op
+    std::array<std::array<Real64, MaxSpectralDataElements>, 5> wlt = {0.0}; // Spectral data wavelengths for each glass layer in a glazing system
```
Probably better off implementing this as an array of array of structs than as seven separate arrays of arrays. I am coming over to this side more and more.

amirroth commented on src/EnergyPlus/WindowManager.hh:
```diff
@@ -804,94 +644,13 @@ struct WindowManagerData : BaseGlobalStruct
-          delta(5, 0.0), hcgap(5, 0.0), hrgap(5, 0.0), rgap(6, 0.0), rs(6, 0.0), arhs(6, 0.0), MaxNumOfIncidentAngles(20),
-          MaxSpectralDataElements(800)
     {
-        AirProps.allocate(8);
```
So good!

Myoldmopar replied:
Yeah!  Like 60 ArrayND objects converted to std arrays.  Hopefully this puts a dent into the number of ArrayND objects where N > 1 so that we can reduce our dependency on Objexx for multidimensional arrays.

amirroth commented on src/EnergyPlus/ZoneDehumidifier.cc:
```diff
@@ -151,22 +151,23 @@ namespace ZoneDehumidifier {
                                       CompName));
             }
-            if (state.dataZoneDehumidifier->CheckEquipName(ZoneDehumidNum)) {
+            if (state.dataZoneDehumidifier->ZoneDehumid(ZoneDehumidNum).CheckEquipName) {
```
Would help to use `auto & zoneDehumid = state.dataZoneDehumidifier->ZoneDehumid(ZoneDehumidNum);` here.

amirroth commented on src/EnergyPlus/ZoneDehumidifier.cc:
```diff
@@ -598,57 +596,36 @@ namespace ZoneDehumidifier {
                 ShowSevereError(state,
-                                "InitZoneDehumidifier: Zone Dehumidifier=\"" + state.dataZoneDehumidifier->ZoneDehumid(LoopIndex).UnitType + ',' +
-                                    state.dataZoneDehumidifier->ZoneDehumid(LoopIndex).Name +
+                                "InitZoneDehumidifier: Zone Dehumidifier=\"" + state.dataZoneDehumidifier->ZoneDehumid(ZoneDehumNum).UnitType + ',' +
```
`format`

amirroth commented on src/EnergyPlus/ZoneDehumidifier.hh:
```diff
@@ -63,142 +63,108 @@ struct EnergyPlusData;
+        Invalid = -1,
+        Discarded, // Default mode where water is "lost"
+        ToTank,    // Collect coil condensate from air and store in water storage tank
+        Num = 2
```
Not necessary to explicitly assign `Num` to 2.

Myoldmopar replied:
Right.  Done.

amirroth commented on src/EnergyPlus/ZoneDehumidifier.hh:
```diff
@@ -63,142 +63,108 @@ struct EnergyPlusData;
+        int AirInletNodeNum = 0;                                              // Inlet air node number
+        int AirOutletNodeNum = 0;                                             // Outlet air node number
+        int WaterRemovalCurveIndex = 0;                                       // Index for water removal curve
+        int WaterRemovalCurveErrorCount = 0;                                  // Count number of times water removal curve returns a negative value
```
This repeating `ErrorCount`, `ErrorIndex` pair pattern should be condensed to a struct at some point.

amirroth commented on src/EnergyPlus/ZoneDehumidifier.hh:
```diff
@@ -63,142 +63,108 @@ struct EnergyPlusData;
-    int CondensateToTank;    // Collect coil condensate from air and store in water storage tank
-
+    bool GetInputFlag = true; // Set to FALSE after first time input is "gotten"
     Array1D<ZoneDehumidifier::ZoneDehumidifierParams> ZoneDehumid;
```
`EPVector`.

amirroth commented on src/EnergyPlus/ZoneEquipmentManager.cc:
```diff
@@ -6362,7 +6362,7 @@ void AutoCalcDOASControlStrategy(EnergyPlusData &state)
+    bool headerAlreadyPrinted = false;
     ErrorsFound = false;
     for (ZoneSizIndex = 1; ZoneSizIndex <= state.dataSize->NumZoneSizingInput; ++ZoneSizIndex) {
         if (state.dataSize->ZoneSizingInput(ZoneSizIndex).AccountForDOAS) {
```
`auto &zoneSizingInput = state.dataSize->ZoneSizingInput(ZoneSizIndex);` (also love shortening `Size` to `Siz`)

amirroth commented on src/EnergyPlus/ZoneEquipmentManager.hh:
```diff
@@ -67,81 +67,73 @@ namespace ZoneEquipmentManager {
-        // Members
         std::string EquipType;
-        int EquipTypeEnum;
+        int EquipTypeEnum = 0;
```
I'll deal with you later, young man!

amirroth commented on src/EnergyPlus/ZoneEquipmentManager.hh:
```diff
@@ -67,81 +67,73 @@ namespace ZoneEquipmentManager {
     void GetZoneEquipment(EnergyPlusData &state);
 
-    void InitZoneEquipment(EnergyPlusData &state, bool const FirstHVACIteration); // unused 1208
+    void InitZoneEquipment(EnergyPlusData &state, bool FirstHVACIteration); // unused 1208
```
Can `FirstHVACIteration` be part of `state` as opposed to a parameter? Asking for a friend.

Myoldmopar replied:
My suggestion would be that we make FirstHVACIteration a part of the HVACManager data structure.  It can certainly live somewhere on state, but I would prefer it not just be in like DataGlobals as one all-encompassing flag.

amirroth replied:
This is obviously not part of this PR, but it's been something that has been bugging me for a while and I just finally blurted it out.

amirroth commented on src/EnergyPlus/ZonePlenum.cc:
```diff
@@ -139,7 +139,7 @@ void SimAirZonePlenum(EnergyPlusData &state,
                            CompName));
             }
-            if (state.dataZonePlenum->CheckRetEquipName(ZonePlenumNum)) {
+            if (state.dataZonePlenum->ZoneRetPlenCond(ZonePlenumNum).checkEquipName) {
```
`auto & zoneRetPlenCond = state.dataZonePlenum->ZoneRetPlenCond(ZonePlenumNum);`

amirroth commented on src/EnergyPlus/ZonePlenum.hh:
```diff
@@ -68,24 +68,23 @@ namespace ZonePlenum {
+        Real64 ZoneTemp = 0.0;
+        Real64 ZoneHumRat = 0.0;
+        Real64 ZoneEnthalpy = 0.0;
+        Real64 OutletTemp = 0.0;
```
This quad also shows up a lot and could be a `struct`.

amirroth commented on src/EnergyPlus/ZoneTempPredictorCorrector.hh:
```diff
@@ -134,74 +134,56 @@ namespace ZoneTempPredictorCorrector {
-        // Members
-        bool initialized;
+        bool initialized = false;
         Array1D<Real64> ThermalComfortAdaptiveASH55_Upper_90;
```
Uhhhh, this can be a `std::array<Array1<Real64>, AdaptiveComfortModel::Num>`.

amirroth commented on src/EnergyPlus/ZoneTempPredictorCorrector.hh:
```diff
@@ -211,36 +193,36 @@ namespace ZoneTempPredictorCorrector {
-    void CalcPredictedHumidityRatio(EnergyPlusData &state, int const ZoneNum, Real64 RAFNFrac);
+    void CalcPredictedHumidityRatio(EnergyPlusData &state, int ZoneNum, Real64 RAFNFrac);
 
     void ReportMoistLoadsZoneMultiplier(Real64 &TotalLoad,
```
Moist!

amirroth commented on src/EnergyPlus/ZoneTempPredictorCorrector.hh:
```diff
@@ -413,7 +389,7 @@ struct ZoneTempPredictorCorrectorData : BaseGlobalStruct
     ZoneTempPredictorCorrector::AdaptiveComfortDailySetPointSchedule AdapComfortDailySetPointSchedule;
 
-    Array1D<Real64> AdapComfortSetPointSummerDesDay;
+    std::array<Real64, 7> AdapComfortSetPointSummerDesDay = {-1};
```
What is 7?

Myoldmopar replied:
It was just what it was allocated to before: `this->AdapComfortSetPointSummerDesDay = Array1D<Real64>(7, -1);`

It looks like it is a comfort model index.  Presumably matching the 7 individual objects you mentioned just [above](https://github.com/NREL/EnergyPlus/pull/9416/files#r868271901).

Myoldmopar committed:
Addressing comments

Myoldmopar commented:
Excellent, OK I think I addressed all the comments critical for this branch.  If CI is happy, this should be able to drop in so we can move on.  Thanks for the comments @amirroth !

Summary:
 - Net reduction of 600 lines of code, primarily in DRYing up all the clear_state functions
 - 93 lines that begin with "Array" were removed, although a few of them were changing Array parameters to span objects
 - Headers should be _much_ easier to read and maintain, with as little impact to the rest of the codebase as possible.

Myoldmopar commented:
The Windows warning there is not related.  The process had trouble deleting a temporary file or something. Mac and Ubuntu are currently starting, so we'll have final regression results here shortly.

Myoldmopar commented:
Solar shading on Linux, unrelated, Mac is happy, all good here and merging.  Thanks again for the review @amirroth
